src/lib/support/utf8.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "utf8.h"
+
+namespace chip {
+namespace Utf8 {
+
+namespace {
+/**
+   State machine for UTF8 valid bytes
+
+Table 3-7. Well-Formed UTF-8 Byte Sequences
+
+Code Points       | First B  | Second B   | Third B | Fourth B
+------------------+----------+------------+---------+---------
+U+0000..U+007F    | 00..7F   |            |         |
+U+0080..U+07FF    | C2..DF   | 80..BF     |         |
+U+0800..U+0FFF    | E0       | A0..BF (A) | 80..BF  |
+U+1000..U+CFFF    | E1..EC   | 80..BF     | 80..BF  |
+U+D000..U+D7FF    | ED       | 80..9F (B) | 80..BF  |
+U+E000..U+FFFF    | EE..EF   | 80..BF     | 80..BF  |
+U+10000..U+3FFFF  | F0       | 90..BF (C) | 80..BF  | 80..BF
+U+40000..U+FFFFF  | F1..F3   | 80..BF     | 80..BF  | 80..BF
+U+100000..U+10FFFF| F4       | 80..8F (D) | 80..BF  | 80..BF
+*/
+
+enum class ParserState
+{
+    kFirstByte,
+    kSecondByte_A,
+    kSecondByte_B,
+    kSecondByte_C,
+    kSecondByte_D,
+    kExtraOneByte,    // 0x80 .. 0xBF  once
+    kExtraTwoBytes,   // 0x80 .. 0xBF  twice
+    kExtraThreeBytes, // 0x80 .. 0xBF  three times
+    //
+    kInvalid, // some error
+};
+
+ParserState NextState(ParserState state, uint8_t value)
+{
+    if (value == 0)
+    {
+        // Refuse to have embedded 0s
+        return ParserState::kInvalid;
+    }
```

Comment:
Embedded zeroes are legal in UTF-8. Let's not mix UTF-8 versus other concerns we have.

Reply:
I documented this. I want this to stay in. This is my chip implementation, not a generic UTF8. Have this in the main comment and unit tests.

Reply:
Added a comment regarding this. This is on purpose. This is why I have CharSpan (not ByteSpan which is UTF8) as well.